components/region_cache_memory_engine/src/engine.rs

```diff
@@ -426,6 +429,21 @@ impl RangeCacheEngine for RangeCacheMemoryEngine {
     type DiskEngine = RocksEngine;
     fn set_disk_engine(&mut self, disk_engine: Self::DiskEngine) {
         self.rocks_engine = Some(disk_engine);
+        let mut bg_work_manager = self.bg_work_manager.write();
+        let self_clone = self.clone();
+        bg_work_manager.set_load_range_and_prepare_for_apply_fn(Arc::new(move |cache_range| {
+            {
+                let mut engine = self_clone.core().write();
+                engine.mut_range_manager().load_range(cache_range.clone())?
+            };
+            Ok(self_clone.prepare_for_apply(cache_range))
```

Comment:
prepare_for_apply should be called when the relevant region begins to apply raft entries. We don't know whether there's some onging writes in the range, so if we schedule the load here (which acquires snapshot right now), ongoing writes may be lost.
For example and range k1 - k10:
In apply thread:        
write k1 in write batch and now the range is not cached nor pending, so k1 will not be buffered in the write batch of memory engine.
In BG worker:
load k1-k10, and schedule load task(acquire snapshot without write k1) and done
In apply thread:  
consume write batch

now, k1 is lost.

Reply:
@tonyxuqqi - I believe we've discussed this issue in our sync on Friday (I've brought it up.) What do you think?

Reply:
@SpadeA-Tang Yes, I also fear this scenario. Any objection if I create an issue to handle it in a follow on PR, along with a test that will replicate it? (Don't want this to grow too large.)

Reply:
> write k1 in write batch and now the range is not cached nor pending, so k1 will not be buffered in the write batch of memory engine.

Writing k1 needs to happen to rocksdb, which generates a new sequence number. Wouldn't we thus not be able to read a version containing k1 as we don't have a snapshot with that sequence number?

Reply:
I think we can handle the issue of not receiving new raft command in the following PR with deleting this line of code.

Reply:
'so i see now: earlier I was buffering all writes in the batch and then making the decision about whether to write them in the callback. this has changed. 

re: moving this to region's apply. 

how would prepare_for_apply know what the relevant range to load is if the cached range is greater than one specific region?  this is the current behavior if we do "ALTER TABLE set attributes" for example. this would require an additional way to communicate that specific range: keep in mind, region_labels can work with labels for an entire table and not just one specific region. 

note that this PR is still fairly limited: in this case, the loading only happens instantly and only when a region is labeled manually via PD. i think it's safe to do it here for this PR.

Reply:
> I think we can handle the issue of not receiving new raft command in the following PR with deleting this line of code.

Done.